### ISSUE TITLE
CLion cannot resolve headers included from other headers

### ISSUE THREAD

**Issue by brownry1986** [bazelbuild/intellij#510]
When opening a header file in CLion that includes a header in a different package, the included header cannot be resolved. The included header is underlined in red and trying to go to declaration produces the "cannot find declaration to go to" message.

![screen shot 2019-02-13 at 5 39 09 pm](https://user-images.githubusercontent.com/6344677/52754563-9154f080-2fc0-11e9-931b-d4ea09e043b1.png)

The project syncs successfully and can be built from the command line with no issues.

I have recreated this in a sample project available here: https://github.com/brownry1986/ClionBazelHeaderSymlinks

Below is my version information:
Operating System: macOS 10.14.2
Bazel Version: 0.22.0-homebrew
CLion Version: 2018.3.4 Build #CL-183.5429.37
Bazel Plugin Version: v2019.01.28.0.2

**Comment by jwnimmer-tri** [comment]
The sample project has `strip_include_prefix = "src/main/include"`.  Does changing that to `includes = ["src/main/include"]` resolve this?  See #145 -- using `strip_include_prefix` did not play well with CLion (and TTBOMK is not yet fixed).

**Comment by brownry1986** [comment]
Indeed, strip_include_prefix does not seem to be supported with CLion. Changing to `includes = ["src/main/include"]` does not seem to have any impact. @jwnimmer-tri I saw that in response to the linked issue you put together an aspect to trick CLion into using a different path to search for includes. Is it possible to use a similar method to add multiple paths for CLion to search for headers?

**Comment by jwnimmer-tri** [comment]
I haven't tried that, but I see no reason why adding several paths would be any worse than just one.  In https://github.com/RobotLocomotion/drake/blob/eacbb43a31d62c1f2b852c90bcef2f907bcf23dc/tools/clion/aspect.bzl#L45 is where I insert the one path; I suspect it could inject several by calling `insert` multiple times.

It's all a bit of a house of cards, though.  If you can live with `includes` and not use `strip_include_prefix` nor `include_prefix`, I would recommend avoiding them.

**Comment by brownry1986** [comment]
@jwnimmer-tri switching to use `includes = [ "src/main/include" ]` did actually work (I didn't give clion enough time to finish loading symbols. However, we don't want to go this route as we don't want to impact how our code is built. Instead we are using an aspect similar to what you referred to above to modify the transitive_include_directory based on the strip_include_prefix and include_prefix. This seems to have fixed our header resolution issues reported here and in #509 . We are still testing out the aspect code, but once it is complete I will post it back here and commit it to the sample project.

**Comment by jin** [comment]
What is the status of the issue? @brownry1986 are you still blocked?

**Comment by brownry1986** [comment]
@jin We have implemented an internal aspect that augments the include info to provide the actual path for include directories instead of the symlinked path. We are planning to post this aspect publicly in the near future.

Is there any appetite to fix this directly in the bazel plugin?

**Comment by siddharthab** [comment]
@brownry1986

How do you handle the case when the #include statement in the source code will need to change to use the real directory structure?

**Comment by brownry1986** [comment]
I have updated the above sample project (https://github.com/brownry1986/ClionBazelHeaderSymlinks) to include a bazel wrapper and an aspect that allows for proper resolution of header files in Clion. In order to use the apsect, you will need to modify the bazel settings in CLion to point the bazel binary path to the location of the bazel wrapper.

The main logic is in the code snippet below where we inspect each transitive include directory and translate the path to modify virtual include paths to the actual path to the include directory.

`def _replace_virtual_includes(transitive_include_directories, ctx):
    results = []
    include_info = _calculate_include_dirs(ctx)

    for t in transitive_include_directories:
        if t.find("/_virtual_includes/") >= 0:
            parts = t.replace("/_virtual_includes/", "/", 1).split("/")
            if parts[BAZEL_OUT_PREFIX_DEPTH] == "external":
                results.append(t)
                continue

            package_name = "/".join(parts[BAZEL_OUT_PREFIX_DEPTH:])

            if package_name in include_info.keys():
                results.extend(include_info[package_name])
            else:
                fail("no include info for package " + package_name)
        else:
            fail("invalid transitive include directory" + t)

    return _uniq(results)
`
